### PR TITLE
Implement telemetry in tracer 

### PR DESCRIPTION
**Pull request by jennchenn**
**What does this PR do?**

This PR is a culmination of all the telemetry work. When telemetry is enabled, `app-started`, `app-integrations-change`, `app-heartbeat` and `app-closing` events should be sent to the telemetry API when applicable.

**Motivation**

Changes have been reviewed incrementally and merged into this feature branch; this PR just merges the feature branch.

**Additional Notes**

**How to test the change?**

Unit tests were added incrementally.

### CODE REVIEW

**Review by delner — APPROVED** [review]
I think this should be good to go!

Just make sure CI is passing first; I think Rubocop updated and broke it. You may need to rebase this PR after `master` is fixed.

File: lib/datadog/core/configuration/settings.rb [hunk 1]
```diff
@@ -618,6 +619,19 @@ def initialize(*_)
           o.default { ENV.fetch(Core::Environment::Ext::ENV_VERSION, nil) }
           o.lazy
         end
+
+        # Client-side telemetry configuration
+        # @public_api
+        settings :telemetry do
+          # Enable telemetry collection. This allows telemetry events to be emitted to the telemetry API.
+          #
+          # @default `DD_INSTRUMENTATION_TELEMETRY_ENABLED` environment variable, otherwise `true`
```
Comment (gingerlime) [inline lib/datadog/core/configuration/settings.rb:628]: The changelog says that "Implement telemetry, disable by default", but this seems like telemetry is enabled by default?

Comment (ivoanjo) [reply lib/datadog/core/configuration/settings.rb:628]: Hey @gingerlime that's actually correct -- we initially merged it enabled (as you see here) but shortly before release we disabled it in https://github.com/DataDog/dd-trace-rb/pull/2233

Hope this helps :)

Comment (gingerlime) [reply lib/datadog/core/configuration/settings.rb:628]: Thank you. The changelog only links to the original PR so I didn't think to check. Perhaps you want to update it? in any case, it's good to know that it's disabled by default!

Comment (ivoanjo) [reply lib/datadog/core/configuration/settings.rb:628]: Ahhh right, oops, you're absolutely right and I've just updated the changelog with the link to the second PR.

Comment (gingerlime) [reply lib/datadog/core/configuration/settings.rb:628]: Thank you @ivoanjo !

**Pull request merged**